File: app/routers/events.py
import json
from datetime import datetime, timezone
from fastapi import APIRouter, Depends
import aiosqlite

from app.database import get_db
from app.models.events import WebEvent, WebEventResponse, CLIEvent, CLIEventResponse
from app.services.slack import notify_prompt_copied, notify_cli_first_run

router = APIRouter()


@router.post("/events", response_model=WebEventResponse)
async def track_event(event: WebEvent, db: aiosqlite.Connection = Depends(get_db)):
    # Upsert user
    now = datetime.now(timezone.utc).isoformat()
    user_type = "web" if event.user_id.startswith("web_") else "cli"
    await db.execute(
        """INSERT INTO users (id, type, created_at, last_seen_at)
           VALUES (?, ?, ?, ?)
           ON CONFLICT(id) DO UPDATE SET last_seen_at = ?""",
        (event.user_id, user_type, now, now, now),
    )

    # Insert event
    cursor = await db.execute(
        "INSERT INTO events (event_type, user_id, session_id, data, created_at) VALUES (?, ?, ?, ?, ?)",
        (event.event_type, event.user_id, event.session_id, json.dumps(event.data) if event.data else None, now),
    )
    await db.commit()

    # Slack notification for copy_prompt
    if event.event_type == "copy_prompt":
        version = (event.data or {}).get("prompt_version", "unknown")
        await notify_prompt_copied(event.session_id or "none", version, event.user_id)

    return WebEventResponse(
        success=True, event_id=cursor.lastrowid, session_id=event.session_id
    )


@router.post("/events/cli", response_model=CLIEventResponse)
async def track_cli_event(event: CLIEvent, db: aiosqlite.Connection = Depends(get_db)):
    now = datetime.now(timezone.utc).isoformat()

    # Check if new user
    row = await db.execute_fetchall("SELECT id FROM users WHERE id = ?", (event.user_id,))
    is_new_user = len(row) == 0

    # Upsert user
    await db.execute(
        """INSERT INTO users (id, type, created_at, last_seen_at)
           VALUES (?, 'cli', ?, ?)
           ON CONFLICT(id) DO UPDATE SET last_seen_at = ?""",
        (event.user_id, now, now, now),
    )

    # Insert CLI event
    await db.execute(
        """INSERT INTO cli_events
           (user_id, session_id, command, profile, duration_ms, success, error_message, cli_version, os, created_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (
            event.user_id, event.session_id, event.command, event.profile,
            event.duration_ms, event.success, event.error_message,
            event.cli_version, event.os, now,
        ),
    )
    await db.commit()

    if is_new_user:
        await notify_cli_first_run(event.user_id, event.command, event.cli_version)

    return CLIEventResponse(success=True)

File: src/handlers/events.rs
use axum::extract::State;
use axum::Json;
use chrono::Utc;
use sqlx::SqlitePool;

use crate::config;
use crate::models::{CLIEvent, CLIEventResponse, WebEvent, WebEventResponse};
use crate::slack;

pub async fn health() -> Json<crate::models::HealthResponse> {
    Json(crate::models::HealthResponse {
        status: "ok".to_string(),
        version: config::APP_VERSION.to_string(),
    })
}

pub async fn track_event(
    State(pool): State<SqlitePool>,
    Json(event): Json<WebEvent>,
) -> Json<WebEventResponse> {
    let now = Utc::now().to_rfc3339();
    let user_type = if event.user_id.starts_with("web_") {
        "web"
    } else {
        "cli"
    };
    let data_json = event
        .data
        .as_ref()
        .map(|d| d.to_string())
        .unwrap_or_default();

    sqlx::query(
        r#"INSERT INTO users (id, type, created_at, last_seen_at)
           VALUES (?, ?, ?, ?)
           ON CONFLICT(id) DO UPDATE SET last_seen_at = ?"#,
    )
    .bind(&event.user_id)
    .bind(user_type)
    .bind(&now)
    .bind(&now)
    .bind(&now)
    .execute(&pool)
    .await
    .ok();

    let row = sqlx::query(
        r#"INSERT INTO events (event_type, user_id, session_id, data, created_at)
           VALUES (?, ?, ?, ?, ?)"#,
    )
    .bind(&event.event_type)
    .bind(&event.user_id)
    .bind(event.session_id.as_deref())
    .bind(if data_json.is_empty() { None::<String> } else { Some(data_json) })
    .bind(&now)
    .execute(&pool)
    .await
    .expect("insert event");

    let event_id = row.last_insert_rowid();

    if event.event_type == "copy_prompt" {
        let version = event
            .data
            .as_ref()
            .and_then(|d| d.get("prompt_version"))
            .and_then(|v| v.as_str())
            .unwrap_or("unknown")
            .to_string();
        let session_id = event.session_id.clone().unwrap_or_else(|| "none".to_string());
        let user_id = event.user_id.clone();
        tokio::spawn(async move {
            slack::notify_prompt_copied(&session_id, &version, &user_id).await;
        });
    }

    Json(WebEventResponse {
        success: true,
        event_id,
        session_id: event.session_id,
    })
}

pub async fn track_cli_event(
    State(pool): State<SqlitePool>,
    Json(event): Json<CLIEvent>,
) -> Json<CLIEventResponse> {
    let now = Utc::now().to_rfc3339();

    let is_new_user: bool = sqlx::query_scalar::<_, String>("SELECT id FROM users WHERE id = ?")
        .bind(&event.user_id)
        .fetch_optional(&pool)
        .await
        .ok()
        .flatten()
        .is_none();

    sqlx::query(
        r#"INSERT INTO users (id, type, created_at, last_seen_at)
           VALUES (?, 'cli', ?, ?)
           ON CONFLICT(id) DO UPDATE SET last_seen_at = ?"#,
    )
    .bind(&event.user_id)
    .bind(&now)
    .bind(&now)
    .bind(&now)
    .execute(&pool)
    .await
    .ok();

    sqlx::query(
        r#"INSERT INTO cli_events
           (user_id, session_id, command, profile, duration_ms, success, error_message, cli_version, os, created_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"#,
    )
    .bind(&event.user_id)
    .bind(event.session_id.as_deref())
    .bind(&event.command)
    .bind(event.profile.as_deref())
    .bind(event.duration_ms)
    .bind(event.success)
    .bind(event.error_message.as_deref())
    .bind(event.cli_version.as_deref())
    .bind(event.os.as_deref())
    .bind(&now)
    .execute(&pool)
    .await
    .expect("insert cli_event");

    if is_new_user {
        let user_id = event.user_id.clone();
        let command = event.command.clone();
        let cli_version = event.cli_version.clone();
        tokio::spawn(async move {
            slack::notify_cli_first_run(&user_id, &command, cli_version.as_deref()).await;
        });
    }

    Json(CLIEventResponse { success: true })
}
